### ISSUE TITLE
Support for JWT in HMAC-SHA256 format

### ISSUE THREAD

**Issue by kholia** [magnumripper/JohnTheRipper#1904]
JSON Web Token (JWT) cracking requires accepting salts of length 137. Currently, the `HMAC-SHA256` format bails out if `salt length > 55` in SIMD mode (which is always enabled?).

**Comment by kholia** [comment]
https://github.com/kholia/JohnTheRipper/tree/jwtbrute has a small script to convert the JWTs into a JtR compatible representation.

**Comment by magnumripper** [comment]
Longer salts are fairly easy to implement.

**Comment by magnumripper** [comment]
Fixed. The exact same changes could be made to the other HMAC formats, if needed.

We currently support 3 limbs (3*64-9 == 183 bytes) of salt but you can bump SALT_LIMBS to an arbitrary number as long as it fits LINE_BUFFER_SIZE...

**Comment by magnumripper** [comment]
So @kholia just add a `prepare()` that converts the JWT hashes to what the format expects!

**Comment by magnumripper** [comment]
BTW our HMAC formats currently handle salts as strings instead of hex blobs. Might be a problem? So perhaps add an alternative format that does hex (or Base64).

Current format is
```
salt_string#hash
```

Perhaps alternate format is simply
```
salt_hex$hash
```
This too should be implemented in all our HMAC formats.

**Comment by jfoug** [comment]
This 'change' has busted the format.   Right now, it fails the TS but only in non-OMP.  NOTE, I am not 100% sure that OMP build actually fixes the problem, but it simply may not show up in the TS run.

If the 4th test string is commented out, then the TS data works just fine. The 4th test is the 137 byte salt item.

So, to me it looks like somewhere we have SIMD buffers not being cleaned up properly (such as the last buffer not getting the 4 bytes of 0's properly set to 0)

**Comment by jfoug** [comment]
NOTE, an ASAN build shows no errors.  Again, this seems to point out clearing failures.  The buffers are properly 'owned', so there is no stray overwriting.

**Comment by jfoug** [comment]
Fixed in 23f77d2 and 2f8a9b1

**Comment by jfoug** [comment]
@kholia I have tried running your script against the sample you provided.  I only get 62 bytes of hash. If I add a character to the end of the sample, I get 64 bytes, but of course, I do not know the 'right' character, NOR the password that should crack it, so can not really test.

**Comment by kholia** [comment]
https://github.com/jmaxxz/jwtbrute

The password seems to be "secret" for the following hash.

`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOjEyMzQ1Njc4OTAsIm5hbWUiOiJKb2huIERvZSIsImFkbWluIjp0cnVlfQ.eoaDVGTClRdfxUZXiPs3f8FmJDkDE_VCQFXqKxpLsts`

**Comment by jfoug** [comment]
Got it.  Your python was not right ;)

**Comment by kholia** [comment]
I don't believe that, you must be lying (LOL).

Thanks @jfoug :+1:

**Comment by jfoug** [comment]
The salt is simply the 'text' up to the 2nd '.' char.  85aa7b3

Fortunately, I found an online hmac calc,and simply started tossing things at it, until one 'hit' with the correct end hash, lol.  Then I found the .cs code, and it appears that also shows the same thing.  The problem in that code, was it split the line on '.' chars, and then rebuilt the first part (with the '.' char), and base-64' the hash.  But the hash you gave me works with the change to split in the format file.
